#### Crawling code

In [1]:
import advertools as adv
import adviz
import pandas as pd
import plotly.express as px
import networkx as nx
from IPython.display import display_markdown
pd.options.display.max_columns = None

def md(text):
    return display_markdown(f'{text}', raw=True)

def make_clickable(val):
    return f'<a href="{val}">{val}</a>'

```python
adv.crawl(
    url_list='https://www.nasa.gov/',
    output_file='nasa_crawl.jl',
    follow_links=True,
    custom_settings={
        'CLOSESPIDER_PAGECOUNT': 1000,
        'LOG_FILE': 'nasa_crawl.log',
        'JOBDIR': 'nasa_crawl',
        'AUTOTHROTTLE_ENABLED': True,
        'AUTOTHROTTLE_TARGET_CONCURRENCY': 6
    })

# convert from .jl to .parquet:
adv.crawlytics.jl_to_parquet('nasa_crawl.jl', 'nasa_crawl.parquet')
```

In [2]:
crawldf = pd.read_parquet('nasa_crawl.parquet')
crawldf.head(3)

url  \
0                                         https://www.nasa.gov/   
1                  https://www.nasa.gov/?search=SpaceX%20Crew-2   
2  https://www.nasa.gov/?search=International%20Space%20Station   

                                                    title  \
0                                                    NASA   
1                 3522 Search Results for "SpaceX Crew-2"   
2  28773 Search Results for "International Space Station"   

                                                                                                                 meta_desc  \
0  NASA.gov brings you the latest news, images and videos from America's space agency, pioneering the future in space e...   
1                                                                                                                     None   
2                                                                                                                     None   

                              viewport charset  \
0  width=device-width, initial-scale=1   UTF-8   
1  width=device-width, initial-scale=1   UTF-8   
2  width=device-width, initial-scale=1   UTF-8   

                                                                                                                        h2  \
0  Suggested Searches@@Martians Wanted@@Featured News@@NASA’s SpaceX Crew-8@@Image Of The Day@@Image Of The Day@@Our Ch...   
1                                                               Suggested Searches@@\n\t\t\t\t3522 results found\t\t\t\t\t   
2                                                              Suggested Searches@@\n\t\t\t\t28773 results found\t\t\t\t\t   

                                                                                                                        h3  \
0  News & Events@@Multimedia@@Featured@@Teams Add Iconic NASA ‘Worm’ Logo to Artemis II Rocket, Spacecraft@@Flame Burns...   
1  News & Events@@Multimedia@@Featured@@Teams Add Iconic NASA ‘Worm’ Logo to Artemis II Rocket, Spacecraft@@Flame Burns...   
2  News & Events@@Multimedia@@Featured@@Teams Add Iconic NASA ‘Worm’ Logo to Artemis II Rocket, Spacecraft@@Flame Burns...   

               canonical  \
0  https://www.nasa.gov/   
1                   None   
2                   None   

                                                                                                                  alt_href  \
0  https://www.nasa.gov/feed/@@https://www.nasa.gov/wp-json/wp/v2/pages/128943@@https://www.nasa.gov/wp-json/oembed/1.0...   
1                                                                                               https://www.nasa.gov/feed/   
2                                                                                               https://www.nasa.gov/feed/   

  og:locale  og:type og:title  \
0     en_US  website     NASA   
1     en_US  website     None   
2     en_US  website     None   

                                                                                                            og:description  \
0  NASA.gov brings you the latest news, images and videos from America's space agency, pioneering the future in space e...   
1                                                                                                                     None   
2                                                                                                                     None   

                  og:url og:site_name            og:updated_time  \
0  https://www.nasa.gov/         NASA  2024-02-17T01:14:00-05:00   
1                   None         NASA                       None   
2                   None         NASA                       None   

                                                                            og:image  \
0  https://www.nasa.gov/wp-content/uploads/2018/07/174116main_2006_01777_highres.jpg   
1                                                                               None   
2                                    

In [3]:
md(f'''
### Rows: {crawldf.shape[0]:,}
### Columns: {crawldf.shape[1]}
--- ''')


### Rows: 9,942
### Columns: 170
--- 

## Link summary table
Maps links on all crawled URLs together with anchor text, rel nofollow attribute, and whether or not it's an internal link.

In [8]:
link_df = adv.crawlytics.links(crawldf, internal_url_regex=r'nasa\.gov')
link_df

url  \
0                                                  https://www.nasa.gov/   
0                                                  https://www.nasa.gov/   
0                                                  https://www.nasa.gov/   
0                                                  https://www.nasa.gov/   
0                                                  https://www.nasa.gov/   
...                                                                  ...   
9941  https://www.nasa.gov/image-detail/bracing-fuel-efficient-flight-2/   
9941  https://www.nasa.gov/image-detail/bracing-fuel-efficient-flight-2/   
9941  https://www.nasa.gov/image-detail/bracing-fuel-efficient-flight-2/   
9941  https://www.nasa.gov/image-detail/bracing-fuel-efficient-flight-2/   
9941  https://www.nasa.gov/image-detail/bracing-fuel-efficient-flight-2/   

                                                                                    link  \
0                                                                  https://www.nasa.gov/   
0                                                                  https://www.nasa.gov/   
0                                                             https://www.nasa.gov/news/   
0                                                    https://www.nasa.gov/news/all-news/   
0                                                          https://plus.nasa.gov/series/   
...                                                                                  ...   
9941                                                                http://oig.nasa.gov/   
9941                                     https://www.nasa.gov/budgets-plans-and-reports/   
9941  https://www.nasa.gov/organizations/budget-annual-reports/agency-financial-reports/   
9941                                                  https://www.nasa.gov/contact-nasa/   
9941                                                 https://www.nasa.gov/accessibility/   

                                                                   text  \
0                                                    \n\t\t\t\t\n\t\t\t   
0                                                    \n\t\t\t\t\n\t\t\t   
0     \n\t\t\t\t\t\t\t\tNews & Events\n\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t\t   
0               \n\t\t\t\t\t\t\t\t\t\tAll NASA News\n\t\t\t\t\t\t\t\t\t   
0       \n\t\t\t\t\t\t\t\t\t\tVideo Series on NASA+\n\t\t\t\t\t\t\t\t\t   
...                                                                 ...   
9941                                                   Office of the IG   
9941                                            Budget & Annual Reports   
9941                                           Agency Financial Reports   
9941                                                       Contact NASA   
9941                                                      Accessibility   

     nofollow  internal  
0       False      True  
0       False      True  
0       False      True  
0       False      True  
0       False      True  
...       ...       ...  
9941    False      True  
9941    False      True  
9941    False      True  
9941    False      True  
9941    False      True  

[1999297 rows x 5 columns]

In [13]:
error_urls = crawldf[crawldf['status'].ne(200)][['url', 'status']].rename(columns={'url': 'link'})
error_urls

link  \
16                                                        https://www.nasa.gov/search/?search=SpaceX%20Crew-2&search_page=177   
17                                       https://www.nasa.gov/search/?search=International%20Space%20Station&search_page=1439   
19                                                    https://www.nasa.gov/search/?search=Mars%20perseverance&search_page=289   
21                                                         https://www.nasa.gov/search/?search=Expedition%2064&search_page=78   
25                                                          https://www.nasa.gov/search/?search=Expedition%2064&search_page=3   
...                                                                                                                       ...   
9837  https://www.nasa.gov/nasa-directorates/armd/advanced-air-vehicles-program/hi-rate-composite-aircraft-manufacturing-p...   
9868                                                             https://www.nasa.gov/jpl/news/ed-stone-colbert-20131204.html   
9869                                                                      https://www.nasa.gov/aeroresearch/stem/seeing-sound   
9870                                                                                               https://www.nasa.gov/qsf18   
9910                          https://www.nasa.gov/feature/goddard/2018/nasa-gold-mission-to-image-earth-s-interface-to-space   

      status  
16       404  
17       404  
19       404  
21       404  
25       404  
...      ...  
9837     404  
9868     404  
9869     404  
9870     404  
9910     404  

[1160 rows x 2 columns]

In [15]:
broken = pd.merge(
    link_df[link_df['link'].isin(error_urls['link'])],
    error_urls,
    left_on='link',
    right_on='link',
    how='left'
)
broken

url  \
0                                                        https://www.nasa.gov/?search=SpaceX%20Crew-2   
1                                        https://www.nasa.gov/?search=International%20Space%20Station   
2                                                    https://www.nasa.gov/?search=Mars%20perseverance   
3                                                        https://www.nasa.gov/?search=Expedition%2064   
4                                                        https://www.nasa.gov/?search=Expedition%2064   
...                                                                                               ...   
4128  https://www.nasa.gov/feature/goddard/2018/nasa-gold-mission-to-image-earth-s-interface-to-space   
4129  https://www.nasa.gov/feature/goddard/2018/nasa-gold-mission-to-image-earth-s-interface-to-space   
4130                                       https://www.nasa.gov/image-article/nasa-going-green-space/   
4131                                       https://www.nasa.gov/image-article/nasa-going-green-space/   
4132     https://www.nasa.gov/missions/icon/nasas-icon-explores-the-boundary-between-earth-and-space/   

                                                                                                 link  \
0                                 https://www.nasa.gov/search/?search=SpaceX%20Crew-2&search_page=177   
1                https://www.nasa.gov/search/?search=International%20Space%20Station&search_page=1439   
2                             https://www.nasa.gov/search/?search=Mars%20perseverance&search_page=289   
3                                   https://www.nasa.gov/search/?search=Expedition%2064&search_page=3   
4                                  https://www.nasa.gov/search/?search=Expedition%2064&search_page=78   
...                                                                                               ...   
4128  https://www.nasa.gov/feature/goddard/2018/nasa-gold-mission-to-image-earth-s-interface-to-space   
4129  https://www.nasa.gov/feature/goddard/2018/nasa-gold-mission-to-image-earth-s-interface-to-space   
4130                                                                      https://www.nasa.gov/spacex   
4131                                                                      https://www.nasa.gov/spacex   
4132                                     https://www.nasa.gov/content/icon-spacecraft-and-instruments   

                                                 text nofollow  internal  \
0                                                 177    False      True   
1                                               1,439    False      True   
2                                                 289    False      True   
3                                                   3    False      True   
4                                                  78    False      True   
...                                               ...      ...       ...   
4128  \n\t\t\t\t\tNews & Events\n\t\t\t\t\t\n\t\t\t\t    False      True   
4129     \n\t\t\t\t\tMultimedia\n\t\t\t\t\t\n\t\t\t\t    False      True   
4130                              technology missions    False      True   
4131                             Space Test Program-2    False      True   
4132                                MIGHTI instrument    False      True   

      status  
0        404  
1        404  
2        404  
3        404  
4        404  
...      ...  
4128     404  
4129     404  
4130     404  
4131     404  
4132     404  

[4133 rows x 6 columns]

## Cleaning up:

* There are many 404 pages linked to from the final pagination links on search results
* There are also many 404s where not-found pages link to themselves. Once we handle these we automatically remove the broken links (to themselves)
* The final broken DataFrame is the same but by filtering out URLs with "search=" and by removing rows where `url` is equal to `link` (a reduction from 4,133 to 1,867 rows)

In [64]:
broken[(~broken['url'].str.contains('search=')) & broken['url'].ne(broken['link'])]

url  \
5                                                                                   https://www.nasa.gov/audio-and-ringtones/   
6                                                                                   https://www.nasa.gov/audio-and-ringtones/   
28    https://www.nasa.gov/centers-and-facilities/ames/nasa-study-reveals-compounding-climate-risks-at-two-degrees-of-warm...   
29    https://www.nasa.gov/centers-and-facilities/ames/nasa-study-reveals-compounding-climate-risks-at-two-degrees-of-warm...   
44                                                                          https://www.nasa.gov/early-stage-innovations-esi/   
...                                                                                                                       ...   
4120                    https://www.nasa.gov/centers-and-facilities/johnson/nasa-goes-quiet-over-galveston-for-flight-series/   
4127                     https://www.nasa.gov/solar-system/why-nasa-watches-airglow-the-colors-of-the-upper-atmospheric-wind/   
4130                                                               https://www.nasa.gov/image-article/nasa-going-green-space/   
4131                                                               https://www.nasa.gov/image-article/nasa-going-green-space/   
4132                             https://www.nasa.gov/missions/icon/nasas-icon-explores-the-boundary-between-earth-and-space/   

                                                                                                 link  \
5                                     https://www.nasa.gov/590319main_ringtone_apollo11_countdown.m4r   
6                                  https://www.nasa.gov/connect/sounds/iphone_install_directions.html   
28                                                                      https://www.nasa.gov/nex/gddp   
29                                                                      https://www.nasa.gov/nex/gddp   
44        https://www.nasa.gov/directorates/spacetech/strg/2012_stro_esi_selections.html#.U3IgSySnw0U   
...                                                                                               ...   
4120                                              https://www.nasa.gov/aeroresearch/stem/seeing-sound   
4127  https://www.nasa.gov/feature/goddard/2018/nasa-gold-mission-to-image-earth-s-interface-to-space   
4130                                                                      https://www.nasa.gov/spacex   
4131                                                                      https://www.nasa.gov/spacex   
4132                                     https://www.nasa.gov/content/icon-spacecraft-and-instruments   

                                                                                       text  \
5                                                             Apollo 11: We Have a Lift-Off   
6     Please visit our iPhone directions for downloading and installing M4R ringtones page.   
28                                                                             found online   
29                                                                   NEX downscaled dataset   
44                                                                                 ESI 2012   
...                                                                                     ...   
4120                                                         “Seeing Sound” learning module   
4127                                                 GOLD Images Earth’s Interface to Space   
4130                                                                    technology missions   
4131                                                                   Space Test Program-2   
4132                                                                      MIGHTI instrument   

     nofollow  internal  status  
5       False      True     404  
6       False      True     404  
28      False      True     404  
29      False      True     404  
4